README.markdown
# Datacomplete

A jQuery UI Autocomplete extension.

[Datacomplete](http://bmaltzan.github.com/jquery-ui-datacomplete)
reads additional data on your page, and sends additional data sent to the server.

![Example](http://bmaltzan.github.com/jquery-ui-datacomplete/static/screenshot.png)



CHANGELOG
jQuery UI Datacomplete Changelog

Version 0.9
==============================================
- Initial release



tests/unit/core.js
module("core");

test("Basic requirements", function() {
expect(1);
ok( $, "$" );
});




src/jquery-ui-datacomplete.0.9.0.js
/*
 * jquery-ui-datacomplete
 * jQuery UI Autocomplete Extension 
 */

(function( $ ) {

var requestIndex = 0;
var itemSelected = 0;

$.widget( "ui.datacomplete", $.ui.autocomplete, {

  options: {
    autoFocus     : true,
    minLength     : 0,
    scrollHeight  : 400,
    extraData     : null,
    columns       : null,
    submitForm    : false,

    select: function(ev, it) {
      var options = $(this).data("datacomplete").options;

      if (options.extraData) {
        for (var i = 0; i < options.extraData.length; i++) {
          var wskey  = options.extraData[i]["ws"];
          var domkey = options.extraData[i]["dom"];
          if (wskey && domkey) {
            domkey = "#" + domkey;

            var ws  = it.item[wskey];
            var dom = $(domkey);

            if (ws && dom) { 
              dom.attr("value", ws);
            }
          }
        }
      }
      itemSelected = true;
    },

                              // form should be submitted as part of close,
                              //   so values will be set and submitted with form.
    close: function() {
      if ( itemSelected ) {
        var options = $(this).data("datacomplete").options;
        if (options.submitForm) {
          var t = $("#" + options.submitForm);
          if (t) {
            t.submit();
          }
        }
      }
    }
  },
                              // this is overridden, instead of just source, 
                              //   so the user can still define urls and use extraParams
  _initSource: function() {
    var self = this,
      array,
      url;
    if ( $.isArray(this.options.source) ) {
      array = this.options.source;
      this.source = function( request, response ) {
        response( $.ui.autocomplete.filter(array, request.term) );
      };
    } else if ( typeof this.options.source === "string" ) {
      url = this.options.source;
      this.source = function( request, response ) {
        if ( self.xhr ) {
          self.xhr.abort();
        }

        if ( this.options.extraParams ) {
          $.each(this.options.extraParams, function(key, param) {
            if ( param == null ) {
              param = document.getElementById(key);
              if ( param && param.length > 0) {
                request[key] = param.value;
              }
            } else {
              request[key] = param;
            }
          });
        }

        self.xhr = $.ajax({
          url: url,
          data: request,
          dataType: "jsonp",
          autocompleteRequest: ++requestIndex,
          success: function( data, status ) {
            if ( this.autocompleteRequest === requestIndex ) {
              response( data );
            }
          },
          error: function() {
            if ( this.autocompleteRequest === requestIndex ) {
              response( [] );
            }
          }
        });
      };
    } else {
      this.source = this.options.source;
    }
  },

  /* When adding divs inside the anchor, need +2 width for Firefox. */
  _resizeMenu: function() {
    var ul = this.menu.element;
    ul.outerWidth( Math.max(
      // Firefox wraps long text (possibly a rounding bug)
      // so we add 1px to avoid the wrapping (#7513)
      ul.width( "" ).outerWidth() + 2,
      this.element.outerWidth()
    ) );
  },

  _renderMenu: function( ul, items ) {
    var self = this;

    if ( this.options.columns && ! items[0].label ) {
      var li = $( "<li></li>" );

      var div = $("<div class='ui-datacomplete-header clearfix'></div>");
      li.append(div);

      for (var i = 0; i < this.options.columns.length; i++) {
        var name = this.options.columns[i];
        for(key in name) {
          var value = name[key];

          var header = $("<div class='dc-header-" + key + "'>" + value +"</div>");
          div.append(header);
        }
      }
      ul.append(li);
    }
    $.each( items, function( index, item ) {
      self._renderItem( ul, item );
    });
  },

  _renderItem: function( ul, item) {
    var li = $( "<li></li>" )
      .data( "item.autocomplete", item );

    var a = $("<a class='clearfix'></a>");
    li.append(a);

    if ( this.options.columns && ! item.label ) {
      for (var i = 0; i < this.options.columns.length; i++) {
        var name = this.options.columns[i];
        for(key in name) {
          var div = $("<div class='dc-" + key + "'>" + item[key] +"</div>");
          a.append(div);
        }
      }
    } else {
			a.append(item.label);
    }

    li.appendTo( ul );
    return li;
  },

                              // item.label is optional now
  _normalize: function( items ) {
    // assume all items have the right format when the first item is complete
    if ( items.length && (this.options.columns || items[0].label) && items[0].value ) {
      return items;
    }
    return $.map( items, function(item) {
      if ( typeof item === "string" ) {
        return {
          label: item,
          value: item
        };
      }
      return $.extend({
        label: item.label || item.value,
        value: item.value || item.label
      }, item );
    });
  },


  testit: function() {
    return "scope test";
  }

});

}( jQuery ));



src/jquery-ui-datacomplete.min.0.9.0.js
(function(b){var h=0,i=0;b.widget("ui.datacomplete",b.ui.autocomplete,{options:{autoFocus:!0,minLength:0,scrollHeight:400,position:{my:"right top",at:"right bottom",collision:"none"},extraData:null,columns:null,submitForm:!1,select:function(a,c){var e=b(this).data("datacomplete").options;if(e.extraData)for(var g=0;g<e.extraData.length;g++){var d=e.extraData[g].ws,f=e.extraData[g].dom;d&&f&&(f="#"+f,d=c.item[d],f=b(f),d&&f&&f.attr("value",d))}i=!0},close:function(){if(i){var a=b(this).data("datacomplete").options;
a.submitForm&&(a=b("#"+a.submitForm))&&a.submit()}}},_initSource:function(){var a=this,c,e;b.isArray(this.options.source)?(c=this.options.source,this.source=function(a,d){d(b.ui.autocomplete.filter(c,a.term))}):typeof this.options.source==="string"?(e=this.options.source,this.source=function(c,d){a.xhr&&a.xhr.abort();this.options.extraParams&&b.each(this.options.extraParams,function(a,b){if(b==null){if((b=document.getElementById(a))&&b.length>0)c[a]=b.value}else c[a]=b});a.xhr=b.ajax({url:e,data:c,
dataType:"json",autocompleteRequest:++h,success:function(a){this.autocompleteRequest===h&&d(a)},error:function(){this.autocompleteRequest===h&&d([])}})}):this.source=this.options.source},_resizeMenu:function(){var a=this.menu.element;a.outerWidth(Math.max(a.width("").outerWidth()+2,this.element.outerWidth()))},_renderMenu:function(a,c){var e=this;if(this.options.columns&&!c[0].label){var g=b("<li></li>"),d=b("<div class='ui-datacomplete-header clearfix'></div>");g.append(d);for(var f=0;f<this.options.columns.length;f++){var j=
this.options.columns[f];for(key in j){var h=b("<div class='dc-header-"+key+"'>"+j[key]+"</div>");d.append(h)}}a.append(g)}b.each(c,function(b,c){e._renderItem(a,c)})},_renderItem:function(a,c){var e=b("<li></li>").data("item.autocomplete",c),g=b("<a class='clearfix'></a>");e.append(g);if(this.options.columns&&!c.label)for(var d=0;d<this.options.columns.length;d++){var f=this.options.columns[d];for(key in f){var h=b("<div class='dc-"+key+"'>"+c[key]+"</div>");g.append(h)}}else g.append(c.label);e.appendTo(a);
return e},_normalize:function(a){if(a.length&&(this.options.columns||a[0].label)&&a[0].value)return a;return b.map(a,function(a){if(typeof a==="string")return{label:a,value:a};return b.extend({label:a.label||a.value,value:a.value||a.label},a)})},testit:function(){return"scope test"}})})(jQuery);


src/jquery-ui-datacomplete.js
/*
 * jquery-ui-datacomplete
 * jQuery UI Autocomplete Extension 
 */

(function( $ ) {

var requestIndex = 0;
var itemSelected = 0;

$.widget( "ui.datacomplete", $.ui.autocomplete, {

  options: {
    autoFocus     : true,
    minLength     : 0,
    scrollHeight  : 400,
    position      : { my        : "right top",
                      at        : "right bottom",
                      collision : "none"
                    },
    extraData     : null,
    columns       : null,
    submitForm    : false,

    select: function(ev, it) {
      var options = $(this).data("datacomplete").options;

      if (options.extraData) {
        for (var i = 0; i < options.extraData.length; i++) {
          var wskey  = options.extraData[i]["ws"];
          var domkey = options.extraData[i]["dom"];
          if (wskey && domkey) {
            domkey = "#" + domkey;

            var ws  = it.item[wskey];
            var dom = $(domkey);

            if (ws && dom) { 
              dom.attr("value", ws);
            }
          }
        }
      }
      itemSelected = true;
    },

                              // form should be submitted as part of close,
                              //   so values will be set and submitted with form.
    close: function() {
      if ( itemSelected ) {
        var options = $(this).data("datacomplete").options;
        if (options.submitForm) {
          var t = $("#" + options.submitForm);
          if (t) {
            t.submit();
          }
        }
      }
    }
  },
                              // this is overridden, instead of just source, 
                              //   so the user can still define urls and use extraParams
  _initSource: function() {
    var self = this,
      array,
      url;
    if ( $.isArray(this.options.source) ) {
      array = this.options.source;
      this.source = function( request, response ) {
        response( $.ui.autocomplete.filter(array, request.term) );
      };
    } else if ( typeof this.options.source === "string" ) {
      url = this.options.source;
      this.source = function( request, response ) {
        if ( self.xhr ) {
          self.xhr.abort();
        }

        if ( this.options.extraParams ) {
          $.each(this.options.extraParams, function(key, param) {
            if ( param == null ) {
              param = document.getElementById(key);
              if ( param && param.length > 0) {
                request[key] = param.value;
              }
            } else {
              request[key] = param;
            }
          });
        }

        self.xhr = $.ajax({
          url: url,
          data: request,
          dataType: "json",
          autocompleteRequest: ++requestIndex,
          success: function( data, status ) {
            if ( this.autocompleteRequest === requestIndex ) {
              response( data );
            }
          },
          error: function() {
            if ( this.autocompleteRequest === requestIndex ) {
              response( [] );
            }
          }
        });
      };
    } else {
      this.source = this.options.source;
    }
  },

  /* When adding divs inside the anchor, need +2 width for Firefox. */
  _resizeMenu: function() {
    var ul = this.menu.element;
    ul.outerWidth( Math.max(
      // Firefox wraps long text (possibly a rounding bug)
      // so we add 1px to avoid the wrapping (#7513)
      ul.width( "" ).outerWidth() + 2,
      this.element.outerWidth()
    ) );
  },

  _renderMenu: function( ul, items ) {
    var self = this;

    if ( this.options.columns && ! items[0].label ) {
      var li = $( "<li></li>" );

      var div = $("<div class='ui-datacomplete-header clearfix'></div>");
      li.append(div);

      for (var i = 0; i < this.options.columns.length; i++) {
        var name = this.options.columns[i];
        for(key in name) {
          var value = name[key];

          var header = $("<div class='dc-header-" + key + "'>" + value +"</div>");
          div.append(header);
        }
      }
      ul.append(li);
    }
    $.each( items, function( index, item ) {
      self._renderItem( ul, item );
    });
  },

  _renderItem: function( ul, item) {
    var li = $( "<li></li>" )
      .data( "item.autocomplete", item );

    var a = $("<a class='clearfix'></a>");
    li.append(a);

    if ( this.options.columns && ! item.label ) {
      for (var i = 0; i < this.options.columns.length; i++) {
        var name = this.options.columns[i];
        for(key in name) {
          var div = $("<div class='dc-" + key + "'>" + item[key] +"</div>");
          a.append(div);
        }
      }
    } else {
			a.append(item.label);
    }

    li.appendTo( ul );
    return li;
  },

                              // item.label is optional now
  _normalize: function( items ) {
    // assume all items have the right format when the first item is complete
    if ( items.length && (this.options.columns || items[0].label) && items[0].value ) {
      return items;
    }
    return $.map( items, function(item) {
      if ( typeof item === "string" ) {
        return {
          label: item,
          value: item
        };
      }
      return $.extend({
        label: item.label || item.value,
        value: item.value || item.label
      }, item );
    });
  },


  testit: function() {
    return "scope test";
  }

});

}( jQuery ));

